lasy/laser.py

```diff
@@ -65,23 +67,132 @@ def __init__(self, dim, lo, hi, npoints, profile,
             # Perform the azimuthal decomposition
             self.field.field[...] = np.fft.ifft(envelope, axis=0)
 
-        normalize_energy(self.dim, profile.laser_energy, self.field)
+    def normalize(self, value, kind=None):
+        """
+        Normalize the pulse either to the energy, peak field amplitude or peak intensity
+
+        Parameters
+        ----------
+        value: scalar
+            Value to which to normalize the field property that is defined in 'kind'
+        kind: string
+            Distance by which the laser pulse should be propagated
+            Options: 'energy', 'field', 'intensity'
+        """
+
+        if kind == 'energy':
+            normalize_energy(self.dim, value, self.field)
+        elif kind == 'field':
+            normalize_peak_field_amplitude(value, self.field)
+        elif kind == 'intensity':
+            normalize_peak_intensity(value, self.field)
+        else:
+            raise ValueError(f'kind "{kind}" not recognized')
 
+    def time_to_frequency(self):
+        """
+        Transform field from the temporal to the frequency domain via FFT,
+        and create the frequency axis if necessary.
+        """
+        times_axis = {'rt': 1, 'xyt': 0}[self.dim]
+        self.field.field_fft = np.fft.fft(self.field.field,
+                                          axis=times_axis,
+                                          norm="forward")
+
+        if not hasattr(self.field, "omega"):
+            dt = self.box.dx[0]
+            omega0 = self.profile.omega0
+            Nt = self.field.field.shape[times_axis]
+            self.field.omega = 2 * np.pi * np.fft.fftfreq(Nt, dt) + omega0
+
+    def frequency_to_time(self):
+        """
+        Transform field from the frequency to the temporal domain via iFFT.
+        """
+        times_axis = {'rt': 1, 'xyt': 0}[self.dim]
+        self.field.field = np.fft.ifft(self.field.field_fft,
+                                       axis=times_axis,
+                                       norm="forward")
 
-    def propagate(self, distance):
+    def move_time_window(self, translate_time):
+        """
+        Translate the `box` and phase of `field_fft` in time by a given amount.
+
+        Parameters
+        ----------
+        translate_time: float (s)
+            Time interval by which the time temporal definitions of the
+            laser should be translated.
+        """
+        self.box.lo[0] += translate_time
+        self.box.hi[0] += translate_time
+        self.box.axes[0] += translate_time
+
+        if self.dim == 'rt':
+            Nt = self.field.field.shape[1]
+            omega_shape = (1, Nt, 1)
+        elif self.dim == 'xyt':
+            Nt = self.field.field.shape[0]
+            omega_shape = (Nt, 1, 1)
```

Comment:
This needs an `else:` branch that aborts

Reply:
that'll be fixed in the pending update of the propagator, please let's leave these routines it as they are for the moment